### PR TITLE
youtrack: 2022.3.65371 -> 2023.3.23390

### PR DESCRIPTION
**Pull request by leona-ya**
## Description of changes

https://www.jetbrains.com/youtrack/whatsnew/

Since JetBrains has deprecated JAR packaging (we used that),
I changed the packaging to use the included ZIP (JetBrains Launcher), renamed the youtrack package to youtrack_2022_3 and add a new package youtrack. 
The update to 2023.x once requires an update with manual interaction. All updates after that will happen automatically.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by leona-ya** [comment]
I'm not really happy with the solution, but I don't have a better one. Also not 100% sure if the configuration options really work that way as the docs confuse me.

File: pkgs/servers/jetbrains/youtrack/default.nix
Comment (dasJ) [inline pkgs/servers/jetbrains/youtrack/default.nix:13]: Can also be `stdenvNoCC` here

File: pkgs/servers/jetbrains/youtrack/default.nix
Comment (dasJ) [inline pkgs/servers/jetbrains/youtrack/default.nix:29]: I haven't tried that myself but wouldn't `makeBinaryWrapper` work as well?

File: pkgs/servers/jetbrains/youtrack/default.nix
Comment (dasJ) [inline pkgs/servers/jetbrains/youtrack/default.nix:45]: Is there anything this fixes? Because I don't set this var and it works perfectly well

File: pkgs/servers/jetbrains/youtrack/default.nix
Comment (dasJ) [inline pkgs/servers/jetbrains/youtrack/default.nix:44]: I only add `gawk` to $PATH, no need to add jdk17 and $out/libexec (which doesn't exist in the first place)

File: pkgs/servers/jetbrains/youtrack/default.nix
Comment (dasJ) [inline pkgs/servers/jetbrains/youtrack/default.nix:47]: I also do `rm -rf $out/internal/java` to get rid of the bundled java

**Review by dasJ — DISMISSED** [review]
Only looked at the package so far

File: pkgs/servers/jetbrains/youtrack/default.nix
Comment (dasJ) [reply pkgs/servers/jetbrains/youtrack/default.nix:29]: Turns out: You don't even need `p7zip` because you already unpacked the zip in the fetcher

Comment (leona-ya) [reply pkgs/servers/jetbrains/youtrack/default.nix:29]: Oh yes! `p7z`ip is actually a relict of a really old derivation i tried.

**Comment by leona-ya** [comment]
Finally I now bumped this to 2023.2 and tested it:
* [x] migration from current nixpkgs master 2022.3 -> this branches 2022.3 works
* [x] migration from 2022.3 -> 2023.2 works (with manual interaction)
* [x] migration within 2023.x works without manual interaction

**Comment by leona-ya** [comment]
I didn't test the explicit youtrack settings, as something seems cursed with that. The easy `services.youtrack.address`, `services.youtrack.port` settings work.

**Comment by leona-ya** [comment]
Updated now to 2023.3

File: nixos/modules/services/web-apps/youtrack.nix
Comment (rvem) [inline nixos/modules/services/web-apps/youtrack.nix:159]: The port is already a part of `settings`
```suggestion
          ${package}/bin/youtrack configure ${lib.concatStringsSep " " (lib.mapAttrsToList (name: value: "--${name}=${toString value}") cfg.settings )}
```

File: nixos/modules/services/web-apps/youtrack.nix
Comment (rvem) [inline nixos/modules/services/web-apps/youtrack.nix:98]: Looks like options from [general parameter](https://www.jetbrains.com/help/youtrack/server/2023.1/YouTrack-Java-Start-Parameters.html#general-parameters) need to be specified as JVM options, i.e. something like `youtrack configure -J-Djetbrains.youtrack.admin.restore=true`, so this particular example doesn't look correct.
Specifying [environment parameters](https://www.jetbrains.com/help/youtrack/server/2023.1/YouTrack-Java-Start-Parameters.html#environmental-parameters) works just fine :relieved:

File: nixos/modules/services/web-apps/youtrack.nix
Comment (rvem) [reply nixos/modules/services/web-apps/youtrack.nix:159]: Also, there is an issue with `youtrack configure` caused by the fact that the config is actually stateful, so for example, you may add `settings.tls-redirect-from-http = true;` to your configuration, switch to the configuration, remove `settings.tls-redirect-from-http = true;` from your configuration, and switch once again. However, `tls-redirect-from-http=1` will still be present in the config files within `/var/lib/youtrack/conf`

Comment (rvem) [reply nixos/modules/services/web-apps/youtrack.nix:159]: Sadly, AFAICS, there is no way to specify these settings as a part of `youtrack run`, so perhaps, it's worth cleaning existing configuration altogether before applying `cfg.settings`

**Review by rvem — COMMENTED** [review]
I was quite surprised to find out that someone already came up with changes required for the ZIP YouTrack distribution :sweat_smile:

File: nixos/modules/services/web-apps/youtrack.nix
Comment (leona-ya) [reply nixos/modules/services/web-apps/youtrack.nix:159]: That is also my level of information. I'm not sure tbh how to clean the configuration? At least AFAIK there is no file that's only written to by `youtrack confiure` and not by the web ui. 
But also the configuration is the one thing that still confuses me, so happy to accept solutions :)

Comment (leona-ya) [reply nixos/modules/services/web-apps/youtrack.nix:159]: But also I'll try that out again

File: nixos/modules/services/web-apps/youtrack.md
Comment (rvem) [inline nixos/modules/services/web-apps/youtrack.md:14]: I had issues with YouTrack DB migration while updating from `2022.3` to `2023.3`.
It was suggested to provide the `-J-Dexodus.entityStore.refactoring.forceAll=true` JVM option for the first YT startup after the upgrade. Perhaps, it's worth mentioning to save someone else time in the future.
However, I couldn't reproduce this on a small YT instance with a free license, so this might only affect some old-ish big YT installations.

File: nixos/modules/services/web-apps/youtrack.nix
Comment (rvem) [reply nixos/modules/services/web-apps/youtrack.nix:159]: > But also the configuration is the one thing that still confuses me, so happy to accept solutions

Agree that the configuration mechanism is now very weird, and I also didn't yet come up with a way to solve this problem:(

Comment (leona-ya) [reply nixos/modules/services/web-apps/youtrack.nix:159]: Just looked at what `youtrack configure` does exactly. When using environment parameters, it sets the value all over the `/conf` directory:
```
[root@nixpkgs:/var/lib/youtrack/conf]# rg listen-port
youtrack/service-config.properties
88:bundle.listen-port=8099

internal/services/adminService/service-config.properties
70:bundle.listen-port=8099

hub/service-config.properties
61:bundle.listen-port=8099

internal/bundle.properties
12:listen-port=8099

internal/services/startingPage/service-config.properties
66:bundle.listen-port=8099

internal/services/bundleProcess/service-config.properties
67:bundle.listen-port=8099

internal/services/configurationWizard/service-config.properties
55:jetbrains.bundle.listen-port.default=8099

internal/services/bundleBackend/service-config.properties
66:bundle.listen-port=8099
``` 

but for jvm options, it just sets them in `/conf/youtrack.jvmoptions`. And this file can probably be stored in the nix store.
```
[root@nixpkgs:/var/lib/youtrack/conf]# rg default.page
youtrack.jvmoptions
65:-Djetbrains.youtrack.default.page=/agiles/157-0/current
```

Comment (leona-ya) [reply nixos/modules/services/web-apps/youtrack.nix]: I now did one approach about that. Not 100% good, but much better than before imo. Feel free to review :)

**Comment by leona-ya** [comment]
Also integrated some more warnings and option migrations now.

File: nixos/modules/services/web-apps/youtrack.nix
Comment (rvem) [reply nixos/modules/services/web-apps/youtrack.nix]: Looks better now:)

**Review by rvem — APPROVED** [review]

**Comment by leona-ya** [comment]
Added one more warning about the deprecation of `services.youtrack.extraParams`.

**Review by kloenk — APPROVED** [review]
LGTM, deployed on my aarch64-linux

**Comment by yu-re-ka** [comment]
do you want to wait until @dasJ re-reviewed, or can I merge this?

**Comment by leona-ya** [comment]
If @dasJ want to review it, I would wait for it :) Otherwise you can probably merge

**Comment by leona-ya** [comment]
I moved the packages to `pkgs/by-name` and removed an unnecessary version check in the module

**Review by kloenk — APPROVED** [review]

**Pull request merged**
